jsystemd-spring-boot-starter/src/main/java/com/github/jpmsilva/jsystemd/SystemdAutoConfiguration.java

```diff
@@ -83,11 +83,13 @@ SystemdNotifyStatusProvider systemdNotifyClassLoaderStatus() {
     @Autowired
     SystemdStatusProviderConfiguration(Systemd systemd, ConfigurableApplicationContext applicationContext,
         ObjectProvider<List<SystemdNotifyStatusProvider>> statuses) {
-      ConfigurableListableBeanFactory beanFactory = applicationContext.getBeanFactory();
-      Set<SystemdNotifyStatusProvider> newProviders = new TreeSet<>(new QuackAnnotationAwareOrderComparator(beanFactory));
```

Comment:
In Java Sets, duplicates are removed if their identity, as defined by equals and hashCode, is the same.
This is in no way related to the order in which they appear, which is a Spring concept.
Elements will not be removed from the set if their order is the same (indeed, the ordering value is not even visible to the TreeSet, only the relative position between elements).
In summary, the TreeSet will not introduce a bug here.

Reply:
From the [Java docs of TreeSet](https://docs.oracle.com/javase/7/docs/api/java/util/TreeSet.html): 
> This is so because the Set interface is defined in terms of the equals operation, but a TreeSet instance performs all element comparisons using its compareTo (or compare) method, so two elements that are deemed equal by this method are, from the standpoint of the set, equal.

So, TreeSet does not use the equals method to compare if two elements are equal, it uses compareTo and if it returns 0 they are deemed equal.

Because when instantiating the TreeSet a new QuackAnnotationAwareOrderComparator was used, when adding elements to the TreeSet it used this Comparator to compare elements in the set and if it returned 0 (this happens if Order is the same) the new element is not added.

I wrote a test: https://gist.github.com/ellull/381b8af7f50eafb0b85f7564739d53b9

Reply:
Ah, I stand firmly corrected, was not aware of that.
Thanks for taking the time to correct my undestanding of TreeSets!